Update Python Conventions Page with Python Hinting

DFriend01 opened the pull request:
<!-- Remember to add the relevant reviewers, assignees, and labels, and create this PR as a draft if it is a work in progress. -->

### Description
<!-- Replace _issue_ with the issue number that this PR resolves, or delete the line and add this PR to the Software project if there is no related issue. -->
- Resolves #102
<!-- Describe what was done in this PR if there is no related issue, or the related issue's description is not sufficient. -->
Added type hinting section in the Python Conventions section.

We should check that type hinting is worth holding as a standard, as the syntax takes some effort to get used to. However, the benefit is that `mypy` allows us to incrementally add type hinting to our code, so our code will not break even if everything does not have type hinting. This would more be for documentation and linting purposes.

DFriend01 committed:
Make python examples at the bottom into collapsable admonitions

DFriend01 committed:
Add type hinting section

patrick-5546 commented on docs/reference/python/conventions.md:
```diff
@@ -36,14 +36,98 @@ and future members of the software team. The major things that we document in ou
 Ideally, the third point should be avoided as much as possible since we would want our code to be
 self explanatory. It should be done only when absolutely necessary.
 
+### Type hinting and static type checking
```
I think this section can be brought out of the documentation section, can put it before. Also, can be renamed to just "Type hinting"

patrick-5546 replied:
This also brings up the discussion of whether we are doing this just for documentation purposes, or for the added benefit of static type checking

patrick-5546 commented on docs/reference/python/conventions.md:
```diff
@@ -36,14 +36,98 @@ and future members of the software team. The major things that we document in ou
+
+Since Python is a dynamically typed language, we need to make use of the [`typing`](https://docs.python.org/3/library/typing.html){target=_blank}
+and [`mypy`](https://mypy.readthedocs.io/en/stable/index.html){target=_blank} modules if we want to do
+any type checking in our functions and classes. Type checking is beneficial for documentation and maintaining a clean
+software architecture. There is some syntax to get familiar in order to use type checking. We recommend the following
```
Explain why we type hint first
- helps catch errors through static type checking
    - mypy is a static type checker

Reference: https://realpython.com/lessons/pros-and-cons-type-hints/

patrick-5546 commented on docs/reference/python/conventions.md:
````diff
@@ -36,14 +36,98 @@ and future members of the software team. The major things that we document in ou
+??? example "Return the sum of a sequence"
+
+    ```python
+    from typing import Sequence, Union
+    Number = Union[int, float]
+
+    def sumseq(seq : Sequence[Number]) -> Number:
+        return sum(seq)
````
Code examples should pass flake8
- 2 spaces between the imports and constants sections, as well as between any classes or global functions

patrick-5546 reviewed (changes requested):
Great work overall, left a couple comments.

One more thing, can you create an infra issue in sailbot_workspace to add mypy to its vscode extensions and tasks, CI, and dev image? Thanks

DFriend01 committed:
Updated python examples to follow flake8 and more information on static type checking

patrick-5546 committed:
Move type hinting section out of documentation and update description

patrick-5546 committed:
Add link to type hinting section

patrick-5546 reviewed (commented):
Should we talk about when to use or not to use type hinting?

DFriend01 committed:
Added cons of type hinting and rules of thumb on when to use it

DFriend01 committed:
Resolved merge conflicts

DFriend01 commented:
I added the cons of type hinting and the rules of thumb from the guide you sent: https://realpython.com/lessons/pros-and-cons-type-hints/

I made sure to cite it.

EDIT: I changed my mind. I just added a link to the guide instead because we didn't add anything unique. May as well just link people to the guide itself.

DFriend01 committed:
Replaced pros and cons with link to guide

patrick-5546 committed:
Fix proofreading regression

patrick-5546 commented on docs/reference/python/conventions.md:
```diff
@@ -18,6 +18,102 @@ pep8 style guide.
     Our CI automatically checks that your code follows the pep8 standard. If it does not, your pull requests will
     be blocked from being merged until those issues are resolved!
 
+### Type hinting
```
Can create 2 sections under style guide:

```md
## Style guide

### Linting

### Type hinting
```

patrick-5546 commented on docs/reference/python/conventions.md:
```diff
@@ -18,6 +18,102 @@ pep8 style guide.
+Expanding on how it catches errors, a static type checker such as [`mypy`](https://mypy.readthedocs.io/en/stable/index.html){target=_blank}
+can be used.
+
+There are [pros and cons](https://realpython.com/lessons/pros-and-cons-type-hints/){target=_blank} to using type
+hinting, make sure that you are aware of them before deciding to utilize them in your codebase.
```
Not sure if this sentence is necessary:

- The link is cited at the bottom of the page
- Target audience is firstly sailbot members, who aren't the ones to decide whether to use type hints

patrick-5546 commented on docs/reference/python/conventions.md:
```diff
@@ -18,6 +18,102 @@ pep8 style guide.
+
+Below are a few examples of using type hinting:
+
+??? example "Return the sum of a sequence"
```
To make it easier to skim read, consider leaving 1 or 2 examples open by default. Can do the same for documentation examples.

DFriend01 committed:
Address latest PR comments

DFriend01 committed:
Proofread
